### PR TITLE
Sort text module

### PR DESCRIPTION
**Pull request by hjorthjort**
This PR moves us closer to the official spec by:
1. Structuring modules fully into separate lists for each kind of definition.
2. Performing the module structuring as part of the transformation from text format to abstract format.

### CODE REVIEW

File: wasm-text.md [hunk 1]
```diff
@@ -369,8 +379,50 @@ Other desugarings are either left for runtime or expressed as macros (for now).
...
+    rule #structureModule((T:TypeDefn   DS:Defns => DS), #module(... types: (TS => T TS)))
+    rule #structureModule((I:ImportDefn DS:Defns => DS), #module(... importDefns: (IS => I IS)))
+    rule #structureModule((X:FuncDefn   DS:Defns => DS), #module(... funcs: (FS => X FS)))
+    rule #structureModule((X:GlobalDefn DS:Defns => DS), #module(... globals: (GS => X GS)))
+    rule #structureModule((T:TableDefn  DS:Defns => DS), #module(... tables: (TS => T TS)))
+    rule #structureModule((M:MemoryDefn DS:Defns => DS), #module(... mems: (MS => M MS)))
+    rule #structureModule((E:ExportDefn DS:Defns => DS), #module(... exports: (ES => E ES)))
+    rule #structureModule((I:DataDefn   DS:Defns => DS), #module(... data: (IS => I IS)))
+    rule #structureModule((I:ElemDefn   DS:Defns => DS), #module(... elem: (IS => I IS)))
+    rule #structureModule((S:StartDefn  DS:Defns => DS), #module(... start: (_ => S .Defns)))
```
Comment (ehildenb) [inline wasm-text.md:412]:
```suggestion
    rule #structureModule((T:TypeDefn   DS:Defns => DS), #module(... types:       TS => T TS))
    rule #structureModule((I:ImportDefn DS:Defns => DS), #module(... importDefns: IS => I IS))
    rule #structureModule((X:FuncDefn   DS:Defns => DS), #module(... funcs:       FS => X FS))
    rule #structureModule((X:GlobalDefn DS:Defns => DS), #module(... globals:     GS => X GS))
    rule #structureModule((T:TableDefn  DS:Defns => DS), #module(... tables:      TS => T TS))
    rule #structureModule((M:MemoryDefn DS:Defns => DS), #module(... mems:        MS => M MS))
    rule #structureModule((E:ExportDefn DS:Defns => DS), #module(... exports:     ES => E ES))
    rule #structureModule((I:DataDefn   DS:Defns => DS), #module(... data:        IS => I IS))
    rule #structureModule((I:ElemDefn   DS:Defns => DS), #module(... elem:        IS => I IS))
    rule #structureModule((S:StartDefn  DS:Defns => DS), #module(... start:       _  => S .Defns))
```

File: wasm-text.md [hunk 2]
```diff
@@ -398,36 +450,48 @@ Record updates can currently not be done in a function rule which also does othe
...
+    syntax Stmt       ::= "#t2aStmt"       "<" Context ">" "(" Stmt       ")" [function]
+    syntax ModuleDecl ::= "#t2aModuleDecl" "<" Context ">" "(" ModuleDecl ")" [function]
+    syntax Defn       ::= "#t2aDefn"       "<" Context ">" "(" Defn       ")" [function]
+    syntax FuncSpec   ::= "#t2aFuncSpec"   "<" Context ">" "(" FuncSpec   ")" [function]
+    syntax TypeUse    ::= "#t2aTypeUse"    "<" Context ">" "(" TypeUse    ")" [function]
+    syntax LocalDecl  ::= "#t2aLocalDecl"  "<" Context ">" "(" LocalDecl  ")" [function]
+ // ------------------------------------------------------------------------------------
     rule text2abstract(DS:Defns) => text2abstract(( module DS ) .Stmts)
-    rule text2abstract(SS)       => #t2aStmts<ctx( ... localIds: .Map)>(unfoldStmts(SS)) [owise]
+    rule text2abstract(SS)       => #t2aStmts<ctx( ... localIds: .Map)>(structureModules(unfoldStmts(SS))) [owise]
```
Comment (ehildenb) [inline wasm-text.md:467]: So here we're mixing `structureModules` directly into the `#t2aStmts` pass.

Does it make sense to instead run `structureModules` across the whole definition, then run `#t2a*`? (similar to how we're running `unfoldDefinitions` before running `#t2a*`?

Comment (hjorthjort) [reply wasm-text.md:467]: Well, it's still one at a time, right? The conversion `text2abstract` first unfolds, then structures, then does `#t2a`. Like function chaingin: `#unfold ; structure ; #t2a`.

Comment (ehildenb) [reply wasm-text.md:467]: Ah, I'm sorry, I was reading it incorrectly, I thought the lhs was #t2a.

**Comment by hjorthjort** [comment]
DeepinScreenshot_select-area_20200605131401
The text format modules look like this. It's the same for all the definition types really, they are just records. Here's functions:
DeepinScreenshot_select-area_20200605131606
In my ideal world, we would not use the parenthesis style at all in the core format, but a more record-like structure.
Using K records make sense for modules, since they have so many many parts, but definitions have so few fields we might as well use a simple strucutre, like `syntax FuncDefn ::= "#func" TypeIdx Valtypes Instrs` for example, or `syntax FuncDefn ::= "#func" "{" TypeIdx "," Valtypes "," Instrs "}"`.
We could also commit to using the parenthesis style in the core format, but we should try to make it unambiguous so that execution of the core format is as simple as possible. And I think it would look nicer if we didn't use the parenthesis style  in the core format then, to ensure there is a clear visible distinction of core format and text format.

**Comment by hjorthjort** [comment]
I'll have a look later today on the formatting changes.

**Comment by ehildenb** [comment]
I like the change to using records, but I think we should be consistent. I don't see any disadvantage to using records for everything.

If you have:

```
syntax Record ::= record ( a: A , b: B , c: C )
```

You can still instantiate it as:

```
rule <k> foo => record(aaa, bbb, ccc) ... </k>
```

without using the record syntax. The record syntax is just a different way to instantiate and pattern match on it.

**Comment by hjorthjort** [comment]
Great, even better.

**Comment by ehildenb** [comment]
But to be clear, the change to record syntax can happen in another PR so that we get the #ContextLookup stuff unblocked.

**Comment by hjorthjort** [comment]
Yes, I was thinking that after the `#ContextLookup` thing is done, every type of definition should be passing through a `#t2a` case, and then we can update them to record form. After that, we can even give the text and abstract format different sorts, which would be nice. But it's not a priority.

**Pull request merged**